### DISCUSSION TITLE
Creating an Interactive Graph Element

### DISCUSSION THREAD

**Discussion by mykolwu** [PrairieLearn/PrairieLearn#7511] [Ideas]
Hi, @ShatteredSk1es and I have been trying to create an interactive graph element that would essentially be a clone of pl-graph, only that it would also let the question writer add a question box on a node or edge for a student to fill out. We were trying to do this using pl-overlay, but were running into issues with positioning the boxes relative to the graph (without simply hard coding the locations for each question).

We're both pretty inexperienced with PL - does anyone have any advice on the best way to create this element, or would be willing to lend us a hand in coding it up?

Thanks so much, I attached a screenshot of what we managed to hardcode so far using pl-graph and pl-overlay.

<img width="335" alt="Screenshot 2023-04-12 at 6 19 22 PM" src="https://user-images.githubusercontent.com/109444053/231620848-de8eba91-87c9-4e51-a762-992302cf71b8.png">

**Comment by eliotwrobson** [comment]
There is a larger effort involving @Surg-dev to create an interactive graph construction element (essentially what is described here, but also allowing repositioning and creating / deleting nodes). Would this meet the needs of this request?

**Comment by nwalters512** [comment]
I'm going to convert this to a discussion - feel free to continue there!

**Comment by mohammadshahnawaz1** [comment]
@Surg-Dev I wanted to ask what progress you've made and what issue's you're having right now. I'm hoping to design a series of questions with an element like this and I'd love to know if you have a working demo. If you don't I'd still like to collab on this if you're down

**Comment by Surg-Dev** [comment]
@mykolwu @ShatteredSk1es 
So we're moving this in steps ATM.

**What's happened so far:**
Right now I'm "working" (I've taken a hiatus for the moment) on a Tree Builder that's been piloted in 225. Basically just clicking to make a root node, then dragging off that root node to make children, able to name them, and submit the tree as-is. Backend grading is custom to whatever the question writer wants (in the case of the piloted question, just insert nodes in BST order). Currently it needs some UX and robustness improvements, which ill get to once my life isn't crazy at the moment.

If you want to test this right now, join the CS225 course instance, it should be a problem at the end of the Exam 3 Practice assignment.

**What's we want to build in the future**
Two levels to this:
1. We want to just extend tree-builder into graph builder, instead of restricting to a tree structure, just build a graph and we grade it (this is basically 2 steps away of being done, as explained later).
2. We want to further have a graph reduction element, where students build pseudocode (?) and a graph visualization appears of what their pseudocode does (??). This is just a far off idea, more applicable to 374/tail end of 225.

**So what did we do?**
So this all stemmed off of when I started working on the DFA-Builder over the summer with Eliot. Basically, there's a public open source one here: https://madebyevan.com/fsm/, we took the code, made it a little bit more robust, embedded it in PL, and hooked it up to the automata package and automatic grading.

We basically took the core of that (since it already builds graphs!) and restricted it down to trees for the tree builder. It would not be hard to rip out DFA logic from it again and just make a generalized element to draw graphs (this weekend, maybe?).

Long story short: **do not use pl elements to make pl elements**. We're using an html canvas and rolling our own graphics code and logic for serialization and submission.

You can look at the evanfsm code, its pretty similar to what we have, you basically just hook into a `<canvas>` element and run a js script from there.

**Comment by mohammadshahnawaz1** [comment]
Ok, I understand, I think our group will definitely try to use and finish the first level of this if we can. Can you send me a link to the CS 225 course instance? I'm not sure what you mean by Exam 3 Practice assignment.

**Reply by Surg-Dev** [reply]
I realize you might not be able to see such an instance, UIUC students can sign up for any stduent-facing course, but non-students will have access to only public-facing. The direct link is [here](https://us.prairielearn.com/pl/course_instance/130633/assessments). But you might have to ask @gcevans for access.

**Reply by mohammadshahnawaz1** [reply]
I see, I'll do that then.

**Comment by mohammadshahnawaz1** [comment]
@gcevans Could you let me and @mykolwu have access to this course instance? We're students at the University of California, Berkeley and are working on this with an end goal of running a study to measure it's effectiveness as an educational tool

**Reply by gcevans** [reply]
I am reaching out to PL to find out what needs to happen to give you access to a course with the question so you can see the current state in action. As to using what we are doing for a study we are really in early days but depending on what your interests are we would be happy to work with you.  Since this work has been largely lead by Sam (@Surg-Dev ) you should work with him as well.

**Reply by Surg-Dev** [reply]
Please email me at samuelr6 (at) illinois.edu, so we can discuss this further!
